micrometer-core/src/main/java/io/micrometer/core/instrument/binder/undertow/UndertowMetrics.java

```diff
@@ -0,0 +1,94 @@
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.TimeGauge;
+import io.micrometer.core.instrument.binder.MeterBinder;
+import io.micrometer.core.lang.NonNullApi;
+import io.micrometer.core.lang.NonNullFields;
+import io.undertow.server.handlers.MetricsHandler;
+import io.undertow.servlet.api.MetricsCollector;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.function.ToDoubleFunction;
+import java.util.function.ToLongFunction;
+
+import static io.undertow.server.handlers.MetricsHandler.MetricResult;
+
+/**
+ * {@link MetricsHandler} to collect metrics for {@link io.undertow.Undertow}
+ *
+ * @author Dharmesh Jogadia
+ * */
+@NonNullApi
+@NonNullFields
+public class UndertowMetrics implements MetricsCollector, MeterBinder {
+
+    private Map<String, MetricsHandler> metricsHandlers;
+    private Iterable<Tag> tags;
+
+    public UndertowMetrics(Iterable<Tag> tags) {
+        metricsHandlers = new HashMap<>();
+        this.tags = tags;
+    }
+
+    @Override
+    public void registerMetric(String servletName, MetricsHandler handler) {
+        metricsHandlers.put(servletName, handler);
+    }
+
+    @Override
+    public void bindTo(MeterRegistry registry) {
+        for (Map.Entry<String, MetricsHandler> handlerEntry : metricsHandlers.entrySet()) {
+            MetricResult metricResult = handlerEntry.getValue().getMetrics();
+            String servletName = handlerEntry.getKey();
+            bindTimer(registry,  "undertow.requests", "Request duration", servletName, metricResult, MetricResult::getTotalRequests, MetricsHandler.MetricResult::getMinRequestTime);
+            bindTimeGauge(registry, "undertow.request.time.max", "Maximum time spent handling requests", servletName, metricResult, MetricResult::getMaxRequestTime);
+            bindTimeGauge(registry, "undertow.request.time.min", "Minimum time spent handling requests", servletName, metricResult, MetricResult::getMinRequestTime);
+            bindCounter(registry,  "undertow.request.errors", "Total number of error requests ", servletName, metricResult, MetricResult::getTotalErrors);
+        }
+    }
+
+    private void bindTimer(MeterRegistry registry, String name, String desc, String servletName, MetricResult metricResult, ToLongFunction<MetricResult> countFunc, ToDoubleFunction<MetricResult> consumer) {
+        FunctionTimer.builder(name, metricResult, countFunc, consumer, TimeUnit.MILLISECONDS)
+                .tags(tags)
+                .tag("servlet_name", servletName)
+                .description(desc)
+                .register(registry);
+    }
+
+    private void bindTimeGauge(MeterRegistry registry, String name, String desc, String servletName, MetricResult metricResult, ToDoubleFunction<MetricResult> consumer) {
+        TimeGauge.builder(name, metricResult, TimeUnit.MILLISECONDS, consumer)
+                .tags(tags)
+                .tag("servlet_name", servletName)
```

Comment:
Tag names should also follow meter naming convention (lowercase dot notation) so they can be properly converted for each metric backend implementation. In this case, `servlet.name`.